### PR TITLE
feat: Support brick/math v0.12

### LINKED ISSUE #525: Update brick/math to support v0.12.0
The brick/math library has updated to [version 0.12.0](https://github.com/brick/math/releases/tag/0.12.0), which includes a few breaking changes in support of the new 8.1-and-above minimum version. Currently this package is locked with 0.11.x as its latest supported version.

Happy to work on a PR to update the dependencies if needed.

### PR DESCRIPTION
**Pull request by siketyan**
Closes #525 

## Description
This pull request adds support for brick/math v0.12.

## Motivation and context
brick/math v0.12 has been released.
ramsey/uuid requires < 0.12 for now, so we can not upgrade brick/math until ramsey/uuid supports it.

## How has this been tested?
Static Analysis + Unit Tests

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: .github/workflows/continuous-integration.yml [hunk 1]
```diff
@@ -79,6 +79,12 @@ jobs:
           - "8.1"
           - "8.2"
           - "8.3"
+        include:
+          # Keep the locked version by default
+          - dependency-versions: "locked"
+          # For PHP 8.0, installing with --prefer-highest to use brick/math v0.11
+          - php-version: "8.0"
+            dependency-versions: "highest"
```
Comment (siketyan) [inline .github/workflows/continuous-integration.yml:87]: Since brick/math is now locked at v0.12.1 (which requires PHP >= 8.1) in composer.lock, CI can not install it on PHP 8.0. Using --prefer-highest to avoid the version and use v0.11. Let me know if there is much better solution.

File: composer.json [hunk 2]
```diff
@@ -11,7 +11,7 @@
     "require": {
         "php": "^8.0",
         "ext-json": "*",
-        "brick/math": "^0.8.8 || ^0.9 || ^0.10 || ^0.11",
+        "brick/math": "^0.8.8 || ^0.9 || ^0.10 || ^0.11 || ^0.12",
```
Comment (siketyan) [inline composer.json:14]: FYI: I tried installing brick/math with --prefer-lowest and run a static analysis, but it failed since BigInteger::toBytes does not exist in v0.8.8. This is out of scope for this PR, so I ignored.

File: psalm.xml [hunk 3]
```diff
@@ -4,7 +4,8 @@
        xsi:schemaLocation="https://getpsalm.org/schema/config vendor/vimeo/psalm/config.xsd"
        errorLevel="1"
        cacheDirectory="./build/cache/psalm"
-       errorBaseline="psalm-baseline.xml">
+       errorBaseline="psalm-baseline.xml"
+       phpVersion="8.1">
```
Comment (siketyan) [inline psalm.xml:8]: Psalm infers the target PHP version from composer.json by default. Enum classes add in brick/math v0.12 are ignored in PHP 8.0, so fixed to PHP 8.1 as the target to avoid it.

File: src/Math/BrickMathCalculator.php [hunk 4]
```diff
@@ -136,9 +136,11 @@ public function toInteger(Hexadecimal $value): IntegerObject
...
     /**
      * Maps ramsey/uuid rounding modes to those used by brick/math
+     *
+     * @return BrickMathRounding::*
      */
-    private function getBrickRoundingMode(int $roundingMode): int
+    private function getBrickRoundingMode(int $roundingMode)
     {
-        return self::ROUNDING_MODE_MAP[$roundingMode] ?? 0;
+        return self::ROUNDING_MODE_MAP[$roundingMode] ?? BrickMathRounding::UNNECESSARY;
```
Comment (siketyan) [inline src/Math/BrickMathCalculator.php:144]: BrickMathRounding::* is a enum case in v0.12 and is a const value in v0.11.

File: .github/workflows/continuous-integration.yml (hunk 1 above)
Comment (baci266) [reply .github/workflows/continuous-integration.yml:87]: maybe drop support for PHP 8.0

https://www.php.net/supported-versions.php

Comment (siketyan) [reply .github/workflows/continuous-integration.yml:87]: I agree, PHP 8.0 is now EOL so we can remove the support.
Let @ramsey decide anyway.

**Review by baci266 — APPROVED** [review]

**Comment by MartinMystikJonas** [comment]
This issue blocks our upgrade because we use another package requiring brick/math 0.12+. Is there anything I can do to help to make this merged?

**Comment by ramsey** [comment]
Sorry for the delay. I've been dealing with some rough issues in my personal life, most notably, I'm unemployed at the moment, so every waking hour is spent job-searching, and I haven't had time for open source contributions. I will try to revisit this and merge and release it as soon as I can, but it might still be a while. Thank you for your patience!

**Comment by nijholt** [comment]
@ramsey any updates on this issue?

**Comment by betterthanclay** [comment]
Hi @ramsey - hope all is well with you. Is there anything we can do to help get this one over the line? It's a showstopper for multiple projects for me that need to be using Brick/Money which requires Brick/Math 0.12+. Thanks!

**Comment by emielmolenaar** [comment]
@ramsey Hope you are doing well. Hoping to merge this one soon. If there is anything we can do, let us know. Thank you.

**Review by ramsey — APPROVED** [review]

**Comment by ramsey** [comment]
Thank you for contributing! 🎉
